### PR TITLE
Adding an optional check for intersections between test and train matrices in predict_rank().

### PR DESCRIPTION
**Pull request by artdgn**
As discussed in[ this issue.](https://github.com/lyst/lightfm/issues/287)

### CODE REVIEW

File: lightfm/lightfm.py [hunk 1]
```diff
@@ -732,8 +732,19 @@ def predict(self, user_ids, item_ids, item_features=None,
 
         return predictions
 
+    def _check_test_train_intersections(self, test_mat, train_mat):
+        if train_mat is not None:
+            bool_intersections_mat = test_mat.astype(bool).multiply(train_mat.astype(bool))
+            n_intersections = np.sum(bool_intersections_mat.tocsr().data)
```
Comment (maciejkula) [inline lightfm/lightfm.py]: Can we not just check `n_intersections = bool_intersections.getnnz()`?

File: lightfm/lightfm.py [hunk 2]
```diff
@@ -732,8 +732,19 @@ def predict(self, user_ids, item_ids, item_features=None,
 
         return predictions
 
+    def _check_test_train_intersections(self, test_mat, train_mat):
+        if train_mat is not None:
+            bool_intersections_mat = test_mat.astype(bool).multiply(train_mat.astype(bool))
```
[the remark this answers is not in the document]
Comment (artdgn) [reply lightfm/lightfm.py]: nope, you're right, it's slightly faster without it.

File: lightfm/lightfm.py (hunk 1 above)
Comment (artdgn) [reply lightfm/lightfm.py]: right again.

**Comment by maciejkula** [comment]
Could you add a way of passing this flag through the evaluation functions that call `predict_rank`?

**Comment by maciejkula** [comment]
This is mostly to maintain a way of getting the old behaviour back. In principle, you're right that it should never be necessary.

**Comment by maciejkula** [comment]
This looks great, thanks for contributing!

**Comment by artdgn** [comment]
Great, thanks!

**Pull request merged**